### PR TITLE
Make addLayer/addSource not generic.

### PR DESCRIPTION
**Pull request by julianrex**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement).

## Pull request checklist:
 - [X] Briefly describe the changes in this PR.
 - [X] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'

### Summary of changes

This PR removes the generic parameter from `addLayer` and `addSource`, instead moving the functionality to a new `StyleEncodable`.

I have not done the same for `getLayer/getSource`, but a similar approach (and a `StyleDecodable`) would work here (though this may be something we don't want to do).

### User impact (optional)

Although these functions have been made non-generic, there is no change to user code.

### CODE REVIEW

File: Sources/MapboxMaps/Style/Style.swift [hunk 1]
```diff
@@ -40,14 +40,11 @@ public class Style {
                 success case. Else, returns a `LayerError` in the `Result` failure case.
      */
     @discardableResult
-    public func addLayer<T: Layer>(layer: T, layerPosition: LayerPosition? = nil) -> Result<Bool, LayerError> {
-
+    public func addLayer(layer: Layer, layerPosition: LayerPosition? = nil) -> Result<Bool, LayerError> {
```
Comment (macdrevx) [inline Sources/MapboxMaps/Style/Style.swift:43]: I want to make these methods throw instead of returning `Result`. Not for this PR though.

Only downside is the errors would be untyped, but I'm not convinced that's a big deal. It would clean up the error handling significantly.

File: Sources/MapboxMaps/Style/StyleErrors.swift [hunk 2]
```diff
@@ -26,7 +30,7 @@ public enum SourceError: Error {
 /// Error enum for all layer-related errors
 public enum LayerError: Error {
     /// The layer provided to the map in `addLayer()` could not be encoded
-    case layerEncodingFailed(Error)
+    case layerEncodingFailed(Error?)
```
Comment (macdrevx) [inline Sources/MapboxMaps/Style/StyleErrors.swift:33]: Why is this change needed?

File: Sources/MapboxMaps/Style/StyleEncodable.swift [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    func jsonObject() throws -> [String: AnyObject]
+}
+
+public extension StyleEncodable where Self: Encodable {
+    /// Given an Encodable object return the JSON dictionary representation
+    /// - Throws: Errors occurring during encoding, or `StyleEncodingError.invalidJSONObject`
+    /// - Returns: A JSON dictionary representing the object.
+    func jsonObject() throws -> [String: AnyObject] {
+        let data = try JSONEncoder().encode(self)
+        guard let jsonObject = try JSONSerialization.jsonObject(with: data) as? [String: AnyObject] else {
```
Comment (macdrevx) [inline Sources/MapboxMaps/Style/StyleEncodable.swift:14]: @nishant-karajgikar are we still planning to refactor this to avoid the serialize/deserialize dance?

**Review by macdrevx — APPROVED** [review]
Nice change

File: Sources/MapboxMaps/Style/Style.swift (hunk 1 above)
Comment (julianrex) [reply Sources/MapboxMaps/Style/Style.swift:43]: Ticketed.

File: Sources/MapboxMaps/Style/StyleErrors.swift (hunk 2 above)
Comment (julianrex) [reply Sources/MapboxMaps/Style/StyleErrors.swift:33]: Ah, test code left-over. Let's revisit as we look at throwing vs Results. I suspect our error cases will change too.
